frontend/components/common/InstantSearchProvider.tsx

```diff
@@ -57,196 +56,207 @@ export function InstantSearchProvider({
-                     path += `/${encodedItemType}`;
-                  }
-               }
-            }
+         const filterObject =
+            typeof filter === 'object' && !Array.isArray(filter) ? filter : {};
 
-            const filterCopy = { ...routeState.filter };
-            if (partsOrTools && deviceHandle) {
-               // Item Type is the slug on device pages, not in the query.
-               delete filterCopy['facet_tags.Item Type'];
+         Object.entries(filterObject).forEach(([attribute, value]) => {
+            if (!facets.includes(attribute)) {
+               delete filterObject[attribute];
+               return;
             }
-            const { q, p, filter, ...otherParams } = qsModule.parse(
-               location.search,
-               {
-                  ignoreQueryPrefix: true,
+            const widgetType = getFacetWidgetType(attribute);
+            switch (widgetType) {
+               case FacetWidgetType.Menu: {
+                  if (isValidRefinementListValue(value)) {
+                     filterObject[attribute] = (value as string[])[0];
+                  } else if (typeof value !== 'string') {
+                     delete filterObject[attribute];
+                  }
+                  break;
                }
-            );
-            const queryString = qsModule.stringify(
-               {
-                  ...otherParams,
-                  ...routeState,
-                  filter: filterCopy,
-               },
-               {
-                  addQueryPrefix: true,
-                  arrayFormat: 'indices',
+               case FacetWidgetType.RefinementList: {
+                  if (!isValidRefinementListValue(value)) {
+                     delete filterObject[attribute];
+                  }
+                  break;
                }
-            );
-
-            return `${baseUrl}${path}${queryString}`;
-         },
-         parseURL({ qsModule, location }) {
-            const pathParts = location.pathname
-               .split('/')
-               .filter((part) => part !== '');
-            const deviceHandle = pathParts.length >= 2 ? pathParts[1] : '';
-            const itemType = pathParts.length >= 3 ? pathParts[2] : '';
+               default: {
+                  return assertNever(widgetType);
+               }
+            }
+         });
 
-            const { q, p, filter } = qsModule.parse(location.search, {
-               ignoreQueryPrefix: true,
-            });
+         if (deviceHandle && itemType) {
+            filterObject['facet_tags.Item Type'] = destylizeDeviceItemType(
+               decodeURIComponent(itemType)
+            ).trim();
+         }
 
-            const filterObject =
-               typeof filter === 'object' && !Array.isArray(filter)
-                  ? filter
-                  : {};
+         return {
+            q: String(q || ''),
+            p: typeof p === 'string' ? parseInt(p) : undefined,
```

Comment:
Dropping the non-negative check on `p` can cause the page to 500, for example visiting `/Parts?p=-1`